include/ginkgo/core/solver/lower_trs.hpp

```diff
@@ -0,0 +1,279 @@
+ * @tparam IndexType  precision of matrix indices
+ *
+ * @ingroup solvers
+ * @ingroup LinOp
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class LowerTrs : public EnableLinOp<LowerTrs<ValueType, IndexType>>,
+                 public Preconditionable {
+    friend class EnableLinOp<LowerTrs>;
+    friend class EnablePolymorphicObject<LowerTrs, LinOp>;
+
+public:
+    using value_type = ValueType;
+    using index_type = IndexType;
+
+    /**
+     * Gets the system operator (matrix) of the linear system.
+     *
+     * @return the system operator (matrix)
+     */
+    std::shared_ptr<const LinOp> get_system_matrix() const
+    {
+        return system_matrix_;
+    }
+
+    /**
+     * Gets the rhs of the linear system.
+     *
+     * @return the rhs
+     */
+    std::shared_ptr<const LinOp> get_rhs() const { return b_; }
+
+    /**
+     * Returns the preconditioner operator used by the solver.
+     *
+     * @return the preconditioner operator used by the solver
+     */
+    std::shared_ptr<const LinOp> get_preconditioner() const override
+    {
+        return preconditioner_;
+    }
+
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        /**
+         * Preconditioner factory.
+         */
+        std::shared_ptr<const LinOpFactory> GKO_FACTORY_PARAMETER(
+            preconditioner, nullptr);
+    };
+#define GKO_COMMA ,
+    GKO_ENABLE_LOWER_TRS_FACTORY(LowerTrs<ValueType GKO_COMMA IndexType>,
+                                 parameters, Factory);
+#undef GKO_COMMA
+    GKO_ENABLE_BUILD_METHOD(Factory);
+
+protected:
+    void apply_impl(const LinOp *b, LinOp *x) const override;
+
+    void apply_impl(const LinOp *alpha, const LinOp *b, const LinOp *beta,
+                    LinOp *x) const override;
+
+    /**
+     * Generates the solver.
+     *
+     * @param system_matrix  the source matrix used to generate the
+     *                      solver.
+     */
+    void generate(const LinOp *system_matrix, const LinOp *b);
+
+    explicit LowerTrs(std::shared_ptr<const Executor> exec)
+        : EnableLinOp<LowerTrs>(std::move(exec))
+    {}
+
+    explicit LowerTrs(const Factory *factory, const LowerTrsArgs &args)
+        : parameters_{factory->get_parameters()},
+          EnableLinOp<LowerTrs>(factory->get_executor(),
+                                transpose(args.system_matrix->get_size())),
+          system_matrix_{std::move(args.system_matrix)},
+          b_{std::move(args.b)}
+    {
+        if (parameters_.preconditioner) {
+            preconditioner_ =
+                parameters_.preconditioner->generate(system_matrix_);
+        } else {
+            preconditioner_ = matrix::Identity<ValueType>::create(
+                this->get_executor(), this->get_size()[0]);
+        }
+        this->generate(gko::lend(system_matrix_), gko::lend(b_));
+    }
+
+private:
+    std::shared_ptr<const LinOp> system_matrix_{};
+    std::shared_ptr<const LinOp> b_{};
```

Comment:
Why are you storing `b_` as a `const LinOp`, when it could be `const Dense<ValueType> instead? You know that it has this types (otherwise, the generation crashes), so stating it makes it more clear (and avoids unnecessary `dynamic_cast`s).

Also: Why do you need to keep the `system_matrix_`? Is there any use for it? I thought you would do everything with the `csr_system_matrix_`, so keeping a pointer to the original `system_matrix_` is unnecessary.